### ISSUE TITLE
flycheck-compile: doesn't properly escape compiler options

### ISSUE THREAD

**Issue by ackalker** [flycheck/flycheck#522]
Trying to help fix #451, I found that using the command `flycheck-compile`, I can get the output of the syntax check in a compilation buffer (nice!). I did encounter the following problem, though:

When there are shell special characters like `(` in one of the strings in the value of `flycheck-clang-definitions`, the command `flycheck-compile` fails with a syntax error. (For example, the Linux kernel uses  the compiler option: `-D"KBUILD_STR(s)=#s"`.)

Here is a simple test case: create a file `test_macro.c` with the following contents:

``` c
#include <stdio.h>

int main(void)
{
  printf("Hello, " KBUILD_STR(NAME) "!\n");
  return 0;
}

/* Local Variables: */
/* flycheck-clang-definitions: ("KBUILD_STR(x)=#x" "NAME=world") */
/* End: */
```

This should syntax check using Flycheck using checker c/c++-clang without errors or warnings.

However, `M-x flycheck-compile <RET> c/c++-clang <RET>`. produces the following error:

```
-*- mode: compilation; default-directory: "~/proj/c/" -*-
Compilation started at Mon Nov 17 17:10:43

clang -fsyntax-only -fno-color-diagnostics -fno-caret-diagnostics -fno-diagnostics-show-option -iquote /home/miki/proj/c/ -Wall -Wextra -DKBUILD_STR(x)=#x -DNAME=world -x c /home/miki/proj/c/test_macro.c
/bin/bash: -c: line 0: syntax error near unexpected token `('
/bin/bash: -c: line 0: `clang -fsyntax-only -fno-color-diagnostics -fno-caret-diagnostics -fno-diagnostics-show-option -iquote /home/miki/proj/c/ -Wall -Wextra -DKBUILD_STR(x)=#x -DNAME=world -x c /home/miki/proj/c/test_macro.c'

Compilation exited abnormally with code 1 at Mon Nov 17 17:10:43
```

It is clear that shell special charaters are not being properly escaped.
I tried quoting (part of) the definition, as well as escaping the special characters, but neither method fixed the error.

Digging deeper, I found the problem to be with the use of `combine-and-quote-strings` in this line in `flycheck-checker-shell-command`: https://github.com/flycheck/flycheck/blob/master/flycheck.el#L2246

From some experimenting, I found that `combine-and-quote-strings` doesn't deal with shell special characters properly: in some cases it puts excess quotes around strings, in others it doesn't properly unescape inside quoted strings:

``` elisp
ELISP> (princ (combine-and-quote-strings '("-DKBUILD_STR(x)=#x" "-DNAME=world")))
-DKBUILD_STR(x)=#x -DNAME=world
"-DKBUILD_STR(x)=#x -DNAME=world"
ELISP> (princ (combine-and-quote-strings '("-D\"KBUILD_STR(x)=#x\"" "-DNAME=world")))
"-D\"KBUILD_STR(x)=#x\"" -DNAME=world
"\"-D\\\"KBUILD_STR(x)=#x\\\"\" -DNAME=world"
ELISP> (princ (combine-and-quote-strings '("-DKBUILD_STR\\(x\\)=\\#x" "-DNAME=world")))
"-DKBUILD_STR\\(x\\)=\\#x" -DNAME=world
"\"-DKBUILD_STR\\\\(x\\\\)=\\\\#x\" -DNAME=world"
```

To work around this, I will make a pull request using `shell-quote-argument` instead of `combine-and-quote-strings`.
